### ISSUE TITLE
Uses of malloc()'ed pointers without NULL check

### ISSUE THREAD

**Issue by solardiz** [openwall/john#5035]
As reported in #5028 and https://huntr.dev/bounties/315d2d54-6ae4-47a1-a6ff-84fea6cb7d01/ by @michaellrowley with assistance of @JamieSlome and @huntr-helper (how many engineers does it take to get a one-liner fix in, and how ridiculous can a bug disclosure process become):

```
NULL Pointer Dereference in openwall/john
Unverified

Reported on

Jan 28th 2022
Description

John The Ripper makes an allocation using malloc() that isn't validated, meaning that - in instances where malloc() returns a null pointer; the null pointer will be dereferenced, typically causing a crash or other undefined behaviour (a system-specific response that is usually something along the lines of a SIGSEGV)
Impact

This vulnerability is capable of allowing attackers to trigger a denial-of-service (or other undefined behaviour) in the event of John The Ripper being unable to access memory (or any other condition where malloc() might fail.)
Occurences

racf_kdfaes_fmt_plug.c L334

This is the original allocation:

unsigned char *t1f = malloc(HASH_OUTPUT_SIZE * cur_salt->mfact);

racf_kdfaes_fmt_plug.c L378

memcpy(t1f + (n*HASH_OUTPUT_SIZE), h_out, HASH_OUTPUT_SIZE);

racf_kdfaes_fmt_plug.c L375

memcpy(m, t1f + (n_key * HASH_OUTPUT_SIZE), HASH_OUTPUT_SIZE);

racf_kdfaes_fmt_plug.c L382-L384

memcpy(t1f + (HASH_OUTPUT_SIZE * (cur_salt->mfact-1)), "\x00\x00\x00\x01", 4);
m1 = (HASH_OUTPUT_SIZE * (cur_salt->mfact-1))+4;
JTR_hmac_sha256(key_p, HASH_OUTPUT_SIZE, t1f, ml, h_out, HASH_OUTPUT_SIZE);

racf_kdfaes_fmt_plug.c L365

memcpy(t1f+(n*HASH_OUTPUT_SIZE), t1, HASH_OUTPUT_SIZE);
```

Of course, this is not a vulnerability, but nevertheless a bug for us to fix. I think proper fix is to replace `malloc` with our `mem_alloc`.

While at it, I also looked into whether `HASH_OUTPUT_SIZE * cur_salt->mfact` can overflow (resulting in UB or/and wrong allocation size, depending on integer types in use). Looks like no: `cur_salt->mfact` is 16-bit unsigned and `HASH_OUTPUT_SIZE` is `32`, so per C promotion rules the 16-bit value gets promoted to `int` first, which on our supported platforms is at least 32-bit. Still, casting to `(size_t)` before multiplication could be a good idea (to explicitly not rely on the promotion rules). That cast would still assume/rely on one of the values being in the 16-bit range individually, so that the multiplication result doesn't overflow even on platforms with 32-bit `size_t`.

**Comment by solardiz** [comment]
Per quick grep, we have some other unchecked uses of `malloc`:

```
7z_common_plug.c:       psalt = malloc(sizeof(sevenzip_salt_t) + cs.aes_length - 1);
7z_common_plug.c-       memcpy(psalt, &cs, sizeof(cs));

dmg2john.c:             chunk1 = (unsigned char *)malloc(data_size);
dmg2john.c-             if (lseek(fd, header2.dataoffset + cno * 4096LL, SEEK_SET) < 0) {
dmg2john.c-                     fprintf(stderr, "Unable to seek in %s\n", filename);
dmg2john.c-                     free(chunk1);
dmg2john.c-                     free(v2_password_header.keyblob);
dmg2john.c-                     goto bailout;
dmg2john.c-             }
dmg2john.c-             count = read(fd, chunk1, data_size);

jumbo.c:                        cp = (char*)malloc(len_now);
jumbo.c-                        len = _vsnprintf(cp, len_now, Fmt, varg);

jumbo.c:        str = malloc(len);
jumbo.c-        sprintf(str, "%s=%s", name, val);

keepass2john.c: buffer = (unsigned char*)malloc(datasize * sizeof(char));
keepass2john.c-
keepass2john.c- /* we inline the content with the hash */
keepass2john.c- fprintf(stderr, "Inlining %s\n", encryptedDatabase);
keepass2john.c- printf("*1*%"PRId64"*", datasize);
keepass2john.c- fseek(fp, 124, SEEK_SET);
keepass2john.c- if (fread(buffer, datasize, 1, fp) != 1) {

and several more like it in keepass2john.c.
```

There's also this, but it might be correct (depending on what's expected from that callback):

```
// Memory allocation callback function
static void* ADL_Main_Memory_Alloc(int iSize)
{
        void*lpBuffer = malloc(iSize);
        return lpBuffer;
}
```

**Comment by solardiz** [comment]
Also with `calloc`:

```
opencl_bitlocker_fmt_plug.c-    padding =
opencl_bitlocker_fmt_plug.c:        (unsigned char *)calloc(BITLOCKER_PADDING_SIZE, sizeof(unsigned char));
opencl_bitlocker_fmt_plug.c-    padding[0] = 0x80;
```

Maybe (is this excluded with the `#ifndef`?):

```
pp.c-#ifndef JTR_MODE
pp.c:  db_entry_t *db_entries   = (db_entry_t *) calloc (pw_max + 1, sizeof (db_entry_t));
pp.c:  pw_order_t *pw_orders    = (pw_order_t *) calloc (pw_max + 1, sizeof (pw_order_t));
pp.c:  u64        *wordlen_dist = (u64 *)        calloc (pw_max + 1, sizeof (u64));
```

**Comment by solardiz** [comment]
Also lots with `strdup` - too many to bother listing here due to a copy-paste pattern. We should probably change those to `str_alloc_copy`.

**Comment by Shashank-Rangu** [comment]
Is it suffice to change the "malloc" in all the places to "mem_alloc"? Or should there be an individual reasoning/explanation for each of these changes with multiple commits?  Is it the same with the strdup--> str_alloc_copy?

**Comment by solardiz** [comment]
For files that are compiled only into `john`, we should change `malloc` to `mem_alloc` and `calloc` to `mem_calloc` - and make sure there's `#include "memory.h"`.

For `*2john`, it can be trickier: depends on whether those already link to `memory.o` or not yet.

For `strdup` to `str_alloc_copy`, it's also trickier: the latter returns memory that can't be freed (is meant to use e.g. on hashes loaded for cracking, and tries to reduce overhead), so it's not a valid universal replacement.

**Comment by solardiz** [comment]
I took a closer look at the uses of `strdup` - no, most of them can't be `str_alloc_copy`. So if we want an easy fix, we probably should introduce an `xstrdup` that would be similar to our `mem_alloc` in terms of error handling, but would wrap `strdup`.

**Comment by solardiz** [comment]
@cr3a7ive-d3v You want to contribute any of these fixes?

**Comment by magnumripper** [comment]
> Maybe (is this excluded with the `#ifndef`?):
> 
> ```
> pp.c-#ifndef JTR_MODE
> pp.c:  db_entry_t *db_entries   = (db_entry_t *) calloc (pw_max + 1, sizeof (db_entry_t));
> pp.c:  pw_order_t *pw_orders    = (pw_order_t *) calloc (pw_max + 1, sizeof (pw_order_t));
> pp.c:  u64        *wordlen_dist = (u64 *)        calloc (pw_max + 1, sizeof (u64));
> ```

Yeah that's always excluded. That file is made so that if JTR_MODE is not defined, the source file is the (logically) pristine PRINCE source code from Jens. That way I can merge his changes with no conflicts, but he didn't change it for years now so I have considered cleaning it away.

Bottom line, that should not be fixed.

**Comment by Shashank-Rangu** [comment]
> @cr3a7ive-d3v You want to contribute any of these fixes?

Yes, I do

**Comment by solardiz** [comment]
@cr3a7ive-d3v Great. Then we'll expect your pull request(s), and hopefully this paves the way for you to contribute in other areas as well. Thank you!

**Comment by solardiz** [comment]
The original issue reported here, along with many other such issues, is now fixed by @Shashank-Rangu in #5054. Now we need a summary of what's left to fix.

**Comment by solardiz** [comment]
> Now we need a summary of what's left to fix.

Nevermind, I am taking care of that in #5059.
